### DISCUSSION TITLE
How to deal with the compensation of PND (particle number density) in PPE matrix and source term with Polygon wall boundary?

### DISCUSSION THREAD

**Discussion by zzqxcy** [rubensamarojr/polymps#8] [General]
the animation seems not good, maybe the compensation work is not finished yet?

**Comment by rubensamarojr** [comment]
Indeed, this compensation has not yet been implemented. 

The Explicitly Represented Polygon (ERP) wall boundary model ([Mitsume et al., 2015](https://doi.org/10.1007/s40571-015-0037-8)) is adopted in the code. This model is fast, as it only requires the closest triangle for all calculations of each partilce, and is accurate for flat planes. However, the lack of neighbors on the corners leads to inaccurate pressures. Notwithstanding, the forces between particle and triangles are OK, as provided in my [work ](https://doi.org/10.1002/fld.5083) using WC-MPS. For the semi-implicit MPS the forces are also OK.

Some work has been proposed to remedy this, although they are more time-consuming:

- Initial layers of particles for all boundary [Zhang et al. 2019](https://doi.org/10.1016/j.compfluid.2018.09.008) or locally [Fourtakas et al. 2019](https://doi.org/10.1016/j.compfluid.2019.06.009)

- Analytical or Semi-analytical boundary integrals [Matsunaga et al. 2020](https://doi.org/10.1002/nme.6429), [Kostorz et al., 2021](https://doi.org/10.1002/nme.6771)
